cirq-rigetti/cirq_rigetti/aspen_device_test.py

```diff
@@ -0,0 +1,181 @@
+import numpy as np
+
+dir_path = pathlib.Path(os.path.dirname(os.path.realpath(__file__)))
+fixture_path = dir_path / '__fixtures__'
+
+
+@pytest.fixture()
+def qcs_aspen8_isa() -> InstructionSetArchitecture:
+    with open(fixture_path / 'QCS-Aspen-8-ISA.json', 'r') as f:
+        return InstructionSetArchitecture.from_dict(json.load(f))
+
+
+def test_octagonal_qubit():
+    """test OctagonalQubit properties"""
+    qubit0 = OctagonalQubit(0)
+    assert qubit0.octagon_position == 0
+    assert qubit0.index == 0
+    assert qubit0.dimension == 2
+    assert OctagonalQubit(1) > qubit0
+
+    qubit5 = OctagonalQubit(5)
+    assert qubit5.x == 0
+    assert qubit5.y == 1 / sqrt(2)
+    assert qubit5.z == 0
+
+    qubit3 = OctagonalQubit(3)
+    assert qubit3.x == 1 + 1 / sqrt(2)
+    assert qubit3.y == 0
+    assert qubit3.z == 0
+
+    qubit2 = OctagonalQubit(2)
+    assert qubit2.x == 1 + sqrt(2)
+    assert qubit2.y == 1 / sqrt(2)
+    assert qubit2.z == 0
+
+    invalid_qubit = OctagonalQubit(0)
+    invalid_qubit.octagon_position = 9
+    with pytest.raises(ValueError):
+        _ = invalid_qubit.x
+    with pytest.raises(ValueError):
+        _ = invalid_qubit.y
+
+    assert np.isclose(qubit0.distance(OctagonalQubit(1)), 1)
+    assert qubit0.distance(OctagonalQubit(7)) == 1
+    with pytest.raises(TypeError):
+        _ = qubit0.distance(AspenQubit(0, 0))
+
+    with pytest.raises(ValueError):
+        _ = OctagonalQubit(8)
+
+
+def test_aspen_qubit():
+    """test AspenQubit properties and transformations"""
+    qubit10 = AspenQubit(1, 0)
+    assert qubit10.octagon == 1
+    assert qubit10.octagon_position == 0
+    assert qubit10.index == 10
+    assert qubit10.dimension == 2
+
+    assert qubit10.x == 3 + 3 / sqrt(2)
+    assert qubit10.y == 1 + sqrt(2)
+    assert np.isclose(qubit10.distance(AspenQubit(0, 7)), 3 + 2 / sqrt(2))
+    assert np.isclose(qubit10.distance(AspenQubit(1, 3)), 1 + sqrt(2))
+
+    qubit15 = AspenQubit(1, 5)
+    assert qubit15.x == (2 + sqrt(2))
+
+    qubit15 = AspenQubit(1, 1)
+    assert qubit15.x == (2 + sqrt(2)) + 1 + sqrt(2)
+
+    invalid_qubit = AspenQubit(0, 0)
+    invalid_qubit.octagon_position = 9
+    with pytest.raises(ValueError):
+        _ = invalid_qubit.x
+    with pytest.raises(ValueError):
+        _ = invalid_qubit.y
+
+    with pytest.raises(TypeError):
+        _ = qubit10.distance(OctagonalQubit(0))
+
+    with pytest.raises(ValueError):
+        _ = AspenQubit(1, 9)
+
+    assert qubit10.to_named_qubit() == cirq.NamedQubit('10')
+    assert AspenQubit.from_named_qubit(cirq.NamedQubit('10')) == AspenQubit(1, 0)
+    with pytest.raises(ValueError):
+        _ = AspenQubit.from_named_qubit(cirq.NamedQubit('s'))
+    with pytest.raises(ValueError):
+        _ = AspenQubit.from_named_qubit(cirq.NamedQubit('19'))
+
+    with pytest.raises(ValueError):
+        _ = qubit10.to_grid_qubit()
+    assert AspenQubit(0, 2).to_grid_qubit() == cirq.GridQubit(0, 0)
```

Comment:
Optional: it might be worth parameterizing this into its own test:

```
@pytest.mark.parametrize(
    'aspen_qubit,cirq_qubit',
    [(AspenQubit(0, 2), cirq.GridQubit(0, 0)),
     ...])
  def test_aspen_qubit_conversion(aspen_qubit,cirq_qubit):
      assert aspen_qubit.to_grid_qubit() == cirq_qubit
      assert aspen_qubit.from_grid_qubit(cirq_qubit) == aspen_qubit
```